bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java

```diff
@@ -77,27 +76,34 @@ public void processBindingConfiguration(String context, Item item, String bindin
 			throw new BindingConfigParseException("context is not permitted to be null for item "+item);
 		}
 
-		Set<Item> items = contextMap.get(context);
-		if (items==null) {
-			items = new HashSet<Item>();
-			contextMap.put(context, items);
+		synchronized (contextMap) {
+			Set<Item> items = contextMap.get(context);
+			if (items == null) {
+				items = new HashSet<Item>();
+				contextMap.put(context, items);
+			}
+
+			items.add(item);
 		}
-			
-		items.add(item);
 	}
 
 	/**
 	 * {@inheritDoc}
 	 */
 	public void removeConfigurations(String context) {
-		Set<Item> items = contextMap.get(context);
-		if(items!=null) {
-			for(Item item : items) {
+		Set<Item> items = null;
+		synchronized (contextMap) {
+			items = contextMap.get(context);
+			if (items != null) {
+				contextMap.remove(context);
+			}
+		}
+		if (items != null) {
+			for (Item item : items) {
 				// we remove all binding configurations for all items
 				bindingConfigs.remove(item.getName());
 				notifyListeners(item);
 			}
-			contextMap.remove(context);
 		}
 	}
 	
```

Comment:
Generally, yes.  That said, I'm not sure how much cross-consistency is expected, by Core, between the `contextMap` and the `bindingConfig` entries.  

The change now has us removing the `bindingConfig` entry _after_ the `contextMap` entry, which reverses the logical ordering from the original code.  I'm not familiar enough with this layer to know if that's a problem or not.